modules/core/src/test/java/org/apache/ignite/internal/processors/datastructures/IgniteReentrantLockTest.java

```diff
@@ -0,0 +1,426 @@
+     * Test a creating of the reentrant lock.
+     *
+     * @throws Exception If failed.
+     */
+    public void testInitialization() throws Exception {
+        try (IgniteLock lock = createReentrantLock(0, "lock1", false)) {
+        }
+
+        try (IgniteLock lock = createReentrantLock(0, "lock2", true)) {
+        }
+    }
+
+    /**
+     * Test a sequential lock acquiring and releasing.
+     *
+     * @throws Exception If failed.
+     */
+    public void testReentrantLock() throws Exception {
+        testReentrantLock(true);
+        testReentrantLock(false);
+    }
+
+    /**
+     * {@link this#testReentrantLock()}
+     *
+     * @throws Exception If failed.
+     */
+    public void testReentrantLock(boolean fair) throws Exception {
+        for (int i = 0; i < NODES_CNT; i++) {
+            final IgniteLock lock = createReentrantLock(i, fair ? "lock1" : "lock2", fair);
+
+            // Other node can still don't see update.
+            GridTestUtils.waitForCondition(new GridAbsPredicate() {
+                @Override public boolean apply() {
+                    return !lock.isLocked();
+                }
+            }, 500L);
+
+            lock.lock();
+
+            assertTrue(lock.isLocked());
+            assertTrue(lock.isHeldByCurrentThread());
+
+            lock.lock();
+
+            assertTrue(lock.isLocked());
+            assertTrue(lock.isHeldByCurrentThread());
+
+            lock.unlock();
+
+            assertTrue(lock.isLocked());
+            assertTrue(lock.isHeldByCurrentThread());
+
+            lock.unlock();
+
+            assertFalse(lock.isLocked());
+            assertFalse(lock.isHeldByCurrentThread());
+        }
+    }
+
+    /**
+     * Test an async lock acquiring and releasing with many nodes.
+     *
+     * @throws Exception If failed.
+     */
+    public void testReentrantLockMultinode() throws Exception {
+        testReentrantLockMultinode(false);
+        testReentrantLockMultinode(true);
+    }
+
+    /**
+     * {@link this#testReentrantLockMultinode()}
+     *
+     * @throws Exception If failed.
+     */
+    public void testReentrantLockMultinode(final boolean fair) throws Exception {
+        final AtomicInteger inx = new AtomicInteger(0);
+
+        GridTestUtils.runMultiThreadedAsync(new Runnable() {
+            @Override public void run() {
+                try {
+                    IgniteLock lock = createReentrantLock(inx.getAndIncrement(), fair ? "lock1" : "lock2", fair);
+                    lock.lock();
+
+                    try {
+                        assertTrue(lock.isLocked());
+                        Thread.sleep(1_000L);
+                    }
+                    catch (InterruptedException e) {
+                        e.printStackTrace();
+                    }
+                    finally {
+                        lock.unlock();
+                    }
+                }
+                catch (Exception e) {
+                    e.printStackTrace();
```

Comment:
mb, its better not to catch exception , just let it go outside. It will be printed anyway

Reply:
Runnable.run() doesn't throw any exceptions.

Reply:
so may be use callable instead

Reply:
fixed